ironfish/src/syncer.ts

```diff
@@ -456,6 +456,15 @@ export class Syncer {
     block: Block
     reason: VerificationResultReason | null
   }> {
+    if (
+      block.header.sequence === GENESIS_BLOCK_SEQUENCE &&
+      peer.genesisBlockHash &&
+      !block.header.hash.equals(peer.genesisBlockHash)
+    ) {
+      peer.punish(BAN_SCORE.MAX, VerificationResultReason.INVALID_GENESIS_BLOCK)
+      return { added: false, block, reason: VerificationResultReason.INVALID_GENESIS_BLOCK }
+    }
```

Comment:
I'm not certain whether this will practically be hit in any code paths right now. This `addBlock` should only be called during either gossip or syncing, and during syncing it's preceded by the ancestor search using `findAncestor`. I just tested out the case where a peer reports a correct genesisHash in their Identify message, but then returns all `0` hashes for block hashes requested in `findAncestor`, and it hits an Assert that throws an error.

So I think findAncestor might be a better place to check for this, but it'd be worth setting up a quick test case locally with a peer that reports a correct genesis block in their Identify message, but then returns incorrect responses later to GetBlockHashes.

Reply:
Changed the check to be in `findAncestor` and also added a check to make sure peers aren't gossiping their genesis block. I'm going to run some additional tests locally tomorrow, but please let me know if the location of these checks seems right.

Reply:
I think this is in the right place -- Just need to retest (manually or otherwise) to verify it's working